gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java

```diff
@@ -845,8 +877,6 @@ public void onContainersAllocated(List<Container> containers) {
               instanceName = null;
             }
           }
-          allocatedContainerCountMap.put(containerHelixTag,
```

Comment:
I re-read the API and I think you're correct. https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-client/apidocs/org/apache/hadoop/yarn/client/api/async/AMRMClientAsync.CallbackHandler.html

Because this method is triggered by heartbeat and this method is called after completion, maybe any implementation would work here since it's not even concurrent lol. All writes to this map actually seem single threaded

Reply:
Another approach is to just add synchronization in `onContainersCompleted`. It doesn't make sense to me that handleCompletion method has no concurrency safe guards and can freely modify the inUsedInstances set while this method is iterating over the set. 

I am hesitant to change too much here because "it works". But at the same time I'd love to make the usage of concurrency primitives consistent so it's not so confusing. 🤷 

Even in the original PR there is a question about usage of explicit lock vs intrinsic lock https://github.com/apache/gobblin/pull/3059

Reply:
Yes, so your change should be better as it add protection when decreasing the container.

Reply:
Turns out it is thread safe to iterate over concurrent hash set. https://stackoverflow.com/questions/56099744/java-concurrenthashset-iterating-over-it-in-a-multi-threaded-environment

And the reason for the lock / synchronized block is because if multiple threads are iterating over the same concurrent hashset and one of the threads calls remove, then since the iterator is essentially a snapshot they could each re-use the same helix instance.

The sync block makes the removal process completely thread safe